README.md

[comment]: # (Start Badges)

[![codecov.io](http://codecov.io/gh/47degrees/pbdirect/branch/master/graph/badge.svg)](http://codecov.io/gh/47degrees/pbdirect) [![Maven Central](https://img.shields.io/badge/maven%20central-0.5.1-green.svg)](https://oss.sonatype.org/#nexus-search;gav~com.47deg~pbdirect*) [![Latest version](https://img.shields.io/badge/pbdirect-0.5.1-green.svg)](https://index.scala-lang.org/47degrees/pbdirect) [![License](https://img.shields.io/badge/License-MIT-blue.svg)](https://raw.githubusercontent.com/47degrees/pbdirect/master/LICENSE) [![Join the chat at https://gitter.im/47deg/pbdirect](https://badges.gitter.im/47deg/pbdirect.svg)](https://gitter.im/47deg/pbdirect?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![GitHub Issues](https://img.shields.io/github/issues/47degrees/pbdirect.svg)](https://github.com/47degrees/pbdirect/issues)

[comment]: # (End Badges)

# PBDirect

Read/Write Scala objects directly to Protobuf with no .proto file definitions

## Context

Protobuf is a fast and efficient way to serialize data. While .proto files are great to share schema definitions between components, it is sometimes much simpler and straightforward to directly encode Scala object without using a .proto schema definition file.

PBDirect aims just that: Make it easier to serialize/deserialize into Protobuf.

## Setup

In order to use PBDirect you need to add the following lines to your `build.sbt`:

[comment]: # (Start Replace)

```scala
libraryDependencies += "com.47deg" %% "pbdirect" % "0.5.2"
```

[comment]: # (End Replace)

## Dependencies

PBDirect depends on:
 - [protobuf-java](https://developers.google.com/protocol-buffers/docs/javatutorial) the Protobuf java library (maintained by Google)
 - [shapeless](https://github.com/milessabin/shapeless) for the generation of type-class instances
 - [cats](https://github.com/typelevel/cats) to deal with optional and repeated fields

## Usage

In order to use PBDirect you need to import the following:

```scala
import pbdirect._
```

## Example

### Schema definition

PBDirect serialises case classes into protobuf and there is no need for a .proto schema definition file.

```scala
case class MyMessage(
  @pbIndex(1) id: Option[Int],
  @pbIndex(3) text: Option[String],
  @pbIndex(5) numbers: List[Int]
)
```

is equivalent to the following protobuf definition:

```protobuf
message MyMessage2 {
  int32  id              = 1;
  string text            = 3;
  repeated int32 numbers = 5;
}
```

Note that the `@pbIndex` annotation is optional. If it is not present, the field's position in the case class is used
as its index. For example, an unannotated case class like:

```scala
case class MyMessage2(
  id: Option[Int],
  text: Option[String],
  numbers: List[Int]
)
```

is equivalent to the following protobuf definition:

```protobuf
message MyMessage {
  int32  id              = 1;
  string text            = 2;
  repeated int32 numbers = 3;
}
```

### Serialization

You only need to call the `toPB` method on your case class. This method is implicitly added with `import pbdirect._`.

```scala
val message = MyMessage2(
  id = Some(123),
  text = Some("Hello"),
  numbers = List(1, 2, 3, 4)
)
// message: MyMessage2 = MyMessage2(Some(123), Some("Hello"), List(1, 2, 3, 4))
val bytes = message.toPB
// bytes: Array[Byte] = Array(
//   8,
//   123,
//   18,
//   5,
//   72,
//   101,
//   108,
//   108,
//   111,
//   26,
//   4,
//   1,
//   2,
//   3,
//   4
// )
```

### Deserialization

Deserializing bytes into a case class is also straight forward. You only need to call the `pbTo[A]` method on the byte array containing the protobuf encoded data.
This method is added implicitly on all `Array[Byte]` by importing `pbdirect._`.

```scala
val bytes2: Array[Byte] = Array[Byte](8, 123, 26, 5, 72, 101, 108, 108, 111, 40, 1, 40, 2, 40, 3, 40, 4)
// bytes2: Array[Byte] = Array(
//   8,
//   123,
//   26,
//   5,
//   72,
//   101,
//   108,
//   108,
//   111,
//   40,
//   1,
//   40,
//   2,
//   40,
//   3,
//   40,
//   4
// )
val message2 = bytes2.pbTo[MyMessage2]
// message2: MyMessage2 = MyMessage2(
//   Some(123),
//   None,
//   List(72, 101, 108, 108, 111)
// )
```

## Extension

You might want to define your own formats for unsupported types.
E.g. to add a format to write `java.time.Instant` you can do:

```scala
import java.time.Instant
import cats.implicits._

implicit val instantFormat: PBFormat[Instant] =
  PBFormat[Long].imap(Instant.ofEpochMilli(_))(_.toEpochMilli)
// instantFormat: PBFormat[Instant] = pbdirect.PBFormat$$anon$1@39639875
```

If you only need a reader you can map over an existing `PBScalarValueReader`

```scala
import java.time.Instant

implicit val instantReader: PBScalarValueReader[Instant] =
  PBScalarValueReader[Long].map(Instant.ofEpochMilli(_))
// instantReader: PBScalarValueReader[Instant] = pbdirect.PBScalarValueReaderImplicits$FunctorReader$$anon$2@4e29fdeb
```

And for a writer you simply contramap over it:

```scala
import java.time.Instant

implicit val instantWriter: PBScalarValueWriter[Instant] =
  PBScalarValueWriter[Long].contramap(_.toEpochMilli)
// instantWriter: PBScalarValueWriter[Instant] = pbdirect.PBScalarValueWriterImplicits$ContravariantWriter$$anon$2@70e50492
```

## Oneof fields

pbdirect supports protobuf [`oneof` fields](https://developers.google.com/protocol-buffers/docs/proto3#oneof) encoded as [Shapeless](https://github.com/milessabin/shapeless) Coproducts.

For example:

```scala
import shapeless._

case class MyMessage3(
  @pbIndex(1) number: Int,
  @pbIndex(2,3,4) coproduct: Option[Int :+: String :+: Boolean :+: CNil]
)
```

is equivalent to the following protobuf definition:

```protobuf
message MyMessage3 {
  int32 number = 1;
  oneof coproduct {
    int32 a  = 2;
    string b = 3;
    bool c   = 4;
  }
}
```

`oneof` fields with exactly two branches can also be encoded using `Either`. For example:

```scala
case class MyMessage4(
  @pbIndex(1) number: Int,
  @pbIndex(2,3) either: Option[Either[String, Boolean]]
)
```

is equivalent to:

```protobuf
message MyMessage4 {
  int32 number = 1;
  oneof either {
    string b = 2;
    bool c   = 3;
  }
}
```

Support for `oneof` fields comes with a couple of restrictions:

* `oneof` fields must have a `@pbIndex` annotation containing the indices of each of the sub-fields
* The type of `oneof` fields must be a Coproduct (or `Either`) wrapped in `Option[_]`. This is so that pbdirect can set the value to `None` when the field is missing when reading a message from protobuf.

## Default values and missing fields

When reading a protobuf message, pbdirect needs to handle missing fields by falling back to some default value. How it does this depends on the type of field.

The following table gives some examples of how pbdirect decodes missing fields:

| Scala type | Value given to missing field |
| --- | --- |
| `Int`/`Short`/`Byte`/`Long` | `0` |
| `Double`/`Float` | `0.0` |
| `String` | `""` |
| `Array[Byte]` | empty array |
| `List[_]` | empty list |
| `Map[_, _]` | empty map |
| Scala `Enumeration` or [enumeratum](https://github.com/lloydmeta/enumeratum) `IntEnum` | the entry with value 0 |
| `Option[_]` | `None` |
| `MyMessage` | an instance of `MyMessage` with all its fields set to their default values |
| `Int :+: String :+: CNil` | (not supported) |
| `Option[Int :+: String :+: CNil]` | `None` |

If you have defined your own `PBScalarValueReader` by mapping over one of the
built-in readers, you will get whatever value is produced by the default value
of the underlying type.

For example, if your message looks like:

```scala
case class MyMessage5(instant: Instant)
```

and you use the `instantReader` defined earlier, reading a message with the `instant` field missing would result in `1970-01-01T00:00:00Z`.

## Packed repeated fields

Primitive repeated fields (ints, floats, doubles, enums and booleans) are
encoded using the protobuf packed encoding by default.

This behaviour can be overriden using the `@pbUnpacked` annotation:

```scala
case class UnpackedMessage(
  @pbUnpacked() ints: List[Int]
)
```

## Fancy integer types (signed/unsigned/fixed-width)

You can tell pbdirect that an Int/Long field should be encoded in a special way
by tagging its type with `Signed`, `Unsigned` or `Fixed`. For example:

```scala
import shapeless.tag.@@
import pbdirect.{Signed, Unsigned, Fixed}

case class IntsMessage(
  normalInt            : Int,
  signedInt            : Int @@ Signed,
  unsignedInt          : Int @@ Unsigned,
  fixedWidthInt        : Int @@ Fixed,
  fixedSignedWidthInt  : Int @@ (Signed with Fixed),
  normalLong           : Long,
  signedLong           : Long @@ Signed,
  unsignedLong         : Long @@ Unsigned,
  fixedWidthLong       : Long @@ Fixed,
  fixedSignedWidthLong : Long @@ (Signed with Fixed)
)
```

would correspond to the following Protobuf definition:


```protobuf
message IntsMessage {
  int32      normalInt              =   1;
  sint32     signedInt              =   2;
  uint32     unsignedInt            =   3;
  fixed32    fixedWidthInt          =   4;
  sfixed32   fixedSignedWidthInt    =   5;
  int64      normalLong             =   6;
  sint64     signedLong             =   7;
  uint64     unsignedLong           =   8;
  fixed64    fixedWidthLong         =   9;
  sfixed64   fixedSignedWidthLong   =   10;
}
```

You can also tag individual types inside coproducts, key and value types of maps
and element types of lists:

```scala
case class AnotherIntsMessage(
  @pbIndex(1, 2) signedIntOrNormalInt  : (Int @@ Signed) :+: Int :+: CNil,
  @pbIndex(3)    signedIntFixedLongMap : Map[Int @@ Signed, Long @@ Fixed],
  @pbIndex(4)    signedIntList         : List[Int @@ Signed]
)
```

[comment]: # (Start Copyright)
# Copyright

pbdirect is designed and developed by 47 Degrees

Copyright (C) 2019-2020 47 Degrees. <http://47deg.com>

[comment]: # (End Copyright)

CHANGELOG.md
# Changelog

## [v0.5.2](https://github.com/47degrees/pbdirect/tree/v0.5.2) (2020-05-08)

[Full Changelog](https://github.com/47degrees/pbdirect/compare/v0.5.1...v0.5.2)

📈 **Dependency updates**

- Update enumeratum to 1.6.0 [\#86](https://github.com/47degrees/pbdirect/pull/86) ([scala-steward](https://github.com/scala-steward))
- Update sbt-ci-release to 1.5.3 [\#83](https://github.com/47degrees/pbdirect/pull/83) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Bumps plugins [\#89](https://github.com/47degrees/pbdirect/pull/89) ([fedefernandez](https://github.com/fedefernandez))
- Update scalafmt-core to 2.5.1 [\#88](https://github.com/47degrees/pbdirect/pull/88) ([BenFradet](https://github.com/BenFradet))
- Deprecate Enum [\#85](https://github.com/47degrees/pbdirect/pull/85) ([cb372](https://github.com/cb372))
- Upgrades Build [\#84](https://github.com/47degrees/pbdirect/pull/84) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Add enum fields to the protoc comparison test [\#66](https://github.com/47degrees/pbdirect/pull/66) ([cb372](https://github.com/cb372))
- Update sbt-org-policies to 0.13.3 [\#65](https://github.com/47degrees/pbdirect/pull/65) ([scala-steward](https://github.com/scala-steward))
- Mergify: configuration update [\#64](https://github.com/47degrees/pbdirect/pull/64) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update sbt-org-policies to 0.13.2 [\#63](https://github.com/47degrees/pbdirect/pull/63) ([scala-steward](https://github.com/scala-steward))

## [v0.5.1](https://github.com/47degrees/pbdirect/tree/v0.5.1) (2020-03-13)

[Full Changelog](https://github.com/47degrees/pbdirect/compare/v0.5.0...v0.5.1)

**Merged pull requests:**

- Release 0.5.1 [\#62](https://github.com/47degrees/pbdirect/pull/62) ([franciscodr](https://github.com/franciscodr))
- Migrates from 47deg to 47degrees GH organization [\#61](https://github.com/47degrees/pbdirect/pull/61) ([franciscodr](https://github.com/franciscodr))
- Update protoc-jar to 3.11.4 [\#60](https://github.com/47degrees/pbdirect/pull/60) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-shapeless\_1.14 to 1.2.5 [\#59](https://github.com/47degrees/pbdirect/pull/59) ([scala-steward](https://github.com/scala-steward))
- Update cats-core to 2.1.1 [\#58](https://github.com/47degrees/pbdirect/pull/58) ([scala-steward](https://github.com/scala-steward))
- Update sbt-org-policies to 0.13.1 [\#57](https://github.com/47degrees/pbdirect/pull/57) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.4.2 [\#56](https://github.com/47degrees/pbdirect/pull/56) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-1-14 to 3.1.1.1 [\#55](https://github.com/47degrees/pbdirect/pull/55) ([scala-steward](https://github.com/scala-steward))

## [v0.5.0](https://github.com/47degrees/pbdirect/tree/v0.5.0) (2020-02-21)

[Full Changelog](https://github.com/47degrees/pbdirect/compare/v0.4.1...v0.5.0)

**Merged pull requests:**

- Fixes openssl command [\#54](https://github.com/47degrees/pbdirect/pull/54) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalacheck-1-14 to 3.1.1.0 [\#52](https://github.com/47degrees/pbdirect/pull/52) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.1 [\#51](https://github.com/47degrees/pbdirect/pull/51) ([scala-steward](https://github.com/scala-steward))
- Releases 0.5.0 [\#49](https://github.com/47degrees/pbdirect/pull/49) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalafmt-core to 2.4.1 [\#48](https://github.com/47degrees/pbdirect/pull/48) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.4.0 [\#47](https://github.com/47degrees/pbdirect/pull/47) ([scala-steward](https://github.com/scala-steward))
- Update protobuf-java to 3.11.4 [\#46](https://github.com/47degrees/pbdirect/pull/46) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.8 [\#45](https://github.com/47degrees/pbdirect/pull/45) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-shapeless\_1.14 to 1.2.4 [\#44](https://github.com/47degrees/pbdirect/pull/44) ([scala-steward](https://github.com/scala-steward))
- Update protobuf-java to 3.11.3 [\#43](https://github.com/47degrees/pbdirect/pull/43) ([scala-steward](https://github.com/scala-steward))
- Update cats-core to 2.1.0 [\#42](https://github.com/47degrees/pbdirect/pull/42) ([scala-steward](https://github.com/scala-steward))
- Cross build for 2.13 and drop 2.11 [\#41](https://github.com/47degrees/pbdirect/pull/41) ([cb372](https://github.com/cb372))
- Update sbt to 1.3.7 [\#39](https://github.com/47degrees/pbdirect/pull/39) ([scala-steward](https://github.com/scala-steward))

## [v0.4.1](https://github.com/47degrees/pbdirect/tree/v0.4.1) (2020-01-13)

[Full Changelog](https://github.com/47degrees/pbdirect/compare/v0.4.0...v0.4.1)

**Merged pull requests:**

- Update sbt-org-policies to 0.12.3 [\#40](https://github.com/47degrees/pbdirect/pull/40) ([scala-steward](https://github.com/scala-steward))
- Release v0.4.1 [\#38](https://github.com/47degrees/pbdirect/pull/38) ([cb372](https://github.com/cb372))
- Support signed/unsigned/fixed-width integers [\#37](https://github.com/47degrees/pbdirect/pull/37) ([cb372](https://github.com/cb372))
- Support reading/writing of Eithers as oneof fields [\#36](https://github.com/47degrees/pbdirect/pull/36) ([cb372](https://github.com/cb372))
- Update enumeratum to 1.5.15 [\#34](https://github.com/47degrees/pbdirect/pull/34) ([scala-steward](https://github.com/scala-steward))

## [v0.4.0](https://github.com/47degrees/pbdirect/tree/v0.4.0) (2020-01-02)

[Full Changelog](https://github.com/47degrees/pbdirect/compare/v0.3.1...v0.4.0)

**Merged pull requests:**

- Release v0.4.0 [\#35](https://github.com/47degrees/pbdirect/pull/35) ([cb372](https://github.com/cb372))

src/main/scala/pbdirect/Pos.scala
package pbdirect

@deprecated("Please use an enumeratum IntEnum instead", since = "0.5.2")
trait Pos {
  val _pos: Int
}

object Pos {
  implicit def posOrdering[P <: Pos]: Ordering[P] =
    new Ordering[P] {
      override def compare(x: P, y: P): Int = x._pos - y._pos
    }

  trait _0  extends Pos { override val _pos = 0  }
  trait _1  extends Pos { override val _pos = 1  }
  trait _2  extends Pos { override val _pos = 2  }
  trait _3  extends Pos { override val _pos = 3  }
  trait _4  extends Pos { override val _pos = 4  }
  trait _5  extends Pos { override val _pos = 5  }
  trait _6  extends Pos { override val _pos = 6  }
  trait _7  extends Pos { override val _pos = 7  }
  trait _8  extends Pos { override val _pos = 8  }
  trait _9  extends Pos { override val _pos = 9  }
  trait _10 extends Pos { override val _pos = 10 }
  trait _11 extends Pos { override val _pos = 11 }
  trait _12 extends Pos { override val _pos = 12 }
  trait _13 extends Pos { override val _pos = 13 }
  trait _14 extends Pos { override val _pos = 14 }
  trait _15 extends Pos { override val _pos = 15 }
  trait _16 extends Pos { override val _pos = 16 }
  trait _17 extends Pos { override val _pos = 17 }
  trait _18 extends Pos { override val _pos = 18 }
  trait _19 extends Pos { override val _pos = 19 }
  trait _20 extends Pos { override val _pos = 20 }
  trait _21 extends Pos { override val _pos = 21 }
  trait _22 extends Pos { override val _pos = 22 }
  trait _23 extends Pos { override val _pos = 23 }
  trait _24 extends Pos { override val _pos = 24 }
  trait _25 extends Pos { override val _pos = 25 }
  trait _26 extends Pos { override val _pos = 26 }
  trait _27 extends Pos { override val _pos = 27 }
  trait _28 extends Pos { override val _pos = 28 }
  trait _29 extends Pos { override val _pos = 29 }
  trait _30 extends Pos { override val _pos = 30 }
  trait _31 extends Pos { override val _pos = 31 }
  trait _32 extends Pos { override val _pos = 32 }
  trait _33 extends Pos { override val _pos = 33 }
  trait _34 extends Pos { override val _pos = 34 }
  trait _35 extends Pos { override val _pos = 35 }
  trait _36 extends Pos { override val _pos = 36 }
  trait _37 extends Pos { override val _pos = 37 }
  trait _38 extends Pos { override val _pos = 38 }
  trait _39 extends Pos { override val _pos = 39 }
  trait _40 extends Pos { override val _pos = 40 }
  trait _41 extends Pos { override val _pos = 41 }
  trait _42 extends Pos { override val _pos = 42 }
  trait _43 extends Pos { override val _pos = 43 }
  trait _44 extends Pos { override val _pos = 44 }
  trait _45 extends Pos { override val _pos = 45 }
  trait _46 extends Pos { override val _pos = 46 }
  trait _47 extends Pos { override val _pos = 47 }
  trait _48 extends Pos { override val _pos = 48 }
  trait _49 extends Pos { override val _pos = 49 }
  trait _50 extends Pos { override val _pos = 50 }
  trait _51 extends Pos { override val _pos = 51 }
  trait _52 extends Pos { override val _pos = 52 }
  trait _53 extends Pos { override val _pos = 53 }
  trait _54 extends Pos { override val _pos = 54 }
  trait _55 extends Pos { override val _pos = 55 }
  trait _56 extends Pos { override val _pos = 56 }
  trait _57 extends Pos { override val _pos = 57 }
  trait _58 extends Pos { override val _pos = 58 }
  trait _59 extends Pos { override val _pos = 59 }
  trait _60 extends Pos { override val _pos = 60 }
  trait _61 extends Pos { override val _pos = 61 }
  trait _62 extends Pos { override val _pos = 62 }
  trait _63 extends Pos { override val _pos = 63 }
  trait _64 extends Pos { override val _pos = 64 }
  trait _65 extends Pos { override val _pos = 65 }
  trait _66 extends Pos { override val _pos = 66 }
  trait _67 extends Pos { override val _pos = 67 }
  trait _68 extends Pos { override val _pos = 68 }
  trait _69 extends Pos { override val _pos = 69 }
  trait _70 extends Pos { override val _pos = 70 }
  trait _71 extends Pos { override val _pos = 71 }
  trait _72 extends Pos { override val _pos = 72 }
  trait _73 extends Pos { override val _pos = 73 }
  trait _74 extends Pos { override val _pos = 74 }
  trait _75 extends Pos { override val _pos = 75 }
  trait _76 extends Pos { override val _pos = 76 }
  trait _77 extends Pos { override val _pos = 77 }
  trait _78 extends Pos { override val _pos = 78 }
  trait _79 extends Pos { override val _pos = 79 }
  trait _80 extends Pos { override val _pos = 80 }
  trait _81 extends Pos { override val _pos = 81 }
  trait _82 extends Pos { override val _pos = 82 }
  trait _83 extends Pos { override val _pos = 83 }
  trait _84 extends Pos { override val _pos = 84 }
  trait _85 extends Pos { override val _pos = 85 }
  trait _86 extends Pos { override val _pos = 86 }
  trait _87 extends Pos { override val _pos = 87 }
  trait _88 extends Pos { override val _pos = 88 }
  trait _89 extends Pos { override val _pos = 89 }
  trait _90 extends Pos { override val _pos = 90 }
  trait _91 extends Pos { override val _pos = 91 }
  trait _92 extends Pos { override val _pos = 92 }
  trait _93 extends Pos { override val _pos = 93 }
  trait _94 extends Pos { override val _pos = 94 }
  trait _95 extends Pos { override val _pos = 95 }
  trait _96 extends Pos { override val _pos = 96 }
  trait _97 extends Pos { override val _pos = 97 }
  trait _98 extends Pos { override val _pos = 98 }
  trait _99 extends Pos { override val _pos = 99 }
}
